### PR TITLE
Add a VSCode .devcontainer

### LINKED ISSUE #1327: Development container for deployment environment
## :white_check_mark: Checklist

- [x] I have searched open and closed issues for duplicates.
- [x] This is a request for a new feature in the Data Safe Haven or an upgrade to an existing feature.
- [x] The feature is still missing in the [latest version](https://github.com/alan-turing-institute/data-safe-haven/releases).
- [x] I have read through the [documentation](https://alan-turing-institute.github.io/data-safe-haven/).
- [x] This isn't an open-ended question (open a [discussion](https://github.com/alan-turing-institute/data-safe-haven/discussions) if it is).

## :strawberry: Suggested change

Microsoft's [Azure TRE](https://microsoft.github.io/AzureTRE/) provides a `.devcontainer` in their Github repository. This integrates nicely with VSCode, automatically creating a container with an appropriate environment for running the deployment scripts. We could add something similar here, which would help avoid issues like those mentioned in #1321 and as mentioned there.

## :steam_locomotive: How could this be done?

Add a `.devcontainer` folder to the repository with an appropriate `Dockerfile` to create a working environment for deployment of the DSH. This would:
- be based on Linux (probably Debian or Ubuntu). 
- be agnostic as to whether the architecture is `amd64` or `arm64`, as either should be fine for deploying the DSH.
- include Powershell and appropriate modules
- include `azure-cli` for interaction with Azure
- (optionally) include `sphinx` and the required dependencies for rebuilding the docs.

I have a [working prototype using Debian](https://github.com/craddm/data-safe-haven/tree/debian-container). Would probably be best to run at least some of the deployment steps from the container (I have tested a few and so far they have worked).

### PR DESCRIPTION
**Pull request by craddm**
## :white_check_mark: Checklist

- [x] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [x] You are targeting the **develop branch**.
- [x] Your branch is up-to-date with the **develop branch** (you probably started your branch from `develop` but it may have changed since then).
- [x] If-and-only-if your changes are not yet ready to merge, you have marked this pull request as a **draft** pull request and added '[WIP]' to the title.
- [ ] If-and-only-if you have changed any Powershell code, you have run the code formatter. You can do this with `./tests/AutoFormat_Powershell.ps1 -TargetPath <path to file or directory>`.

### :arrow_heading_up: Summary

Adds a `.devcontainer` folder to the repo. For VSCode users, this will automatically create and run a Docker container based on Debian 10 with the right dependencies installed for deployment of the DSH (Powershell, azure-cli). The Dockerfile included may be used separately with Docker or podman to create the container, if desired. 

### :closed_umbrella: Related issues

Closes #1327 

### :microscope: Tests

Still to be done is a full deployment from the container - have run some steps successfully but may be good to check if full deployment works as expected.

### CODE REVIEW

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+# RUN apt-get update \
+#   && apt-get -y install ca-certificates curl \ 
+#   apt-transport-https lsb-release gnupg \
+#   && curl -sL https://packages.microsoft.com/keys/microsoft.asc \ 
+#   | gpg --dearmor | tee /etc/apt/trusted.gpg.d/microsoft.gpg > /dev/null \
+#   && AZ_REPO="$(lsb_release -cs)" \
+#   && echo "deb [arch=$(dpkg --print-architecture)] https://packages.microsoft.com/repos/azure-cli/ $AZ_REPO main" \ 
+#   | tee /etc/apt/sources.list.d/azure-cli.list \
+#   && apt-get update \
+#   && apt-get -y install azure-cli="${AZURE_CLI_VERSION}-1~buster"
```
Comment (jemrobinson) [inline .devcontainer/Dockerfile]: Are these lines needed?

File: .devcontainer/install_pwsh_modules.ps1 [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    "Powershell-Yaml"                = @("ge", "0.4.2")
+    "PSScriptAnalyzer"               = @("ge", "1.21.0")
+}
+
+# Powershell modules
+$RepositoryName = "PSGallery"
+Set-PSRepository -Name $RepositoryName -InstallationPolicy Trusted
+foreach ($ModuleName in $ModuleVersionRequired.Keys) {
+    Install-Module -Name $ModuleName -RequiredVersion $RequiredVersion -Repository $RepositoryName -Scope CurrentUser
+}
```
Comment (jemrobinson) [inline .devcontainer/install_pwsh_modules.ps1]: This looks like a duplicate of the existing installation script. Is there a way we can define this all in one place? For example by copying the existing script into the devcontainer?

File: .devcontainer/requirements.txt [hunk 3]
```diff
@@ -0,0 +1,9 @@
+emoji==2.1.0
+GitPython==3.1.28
+Jinja2==3.1.2
+myst-parser==0.18.1
+Pygments==2.13.0
+pydata-sphinx-theme==0.11.0
+rinohtype==0.5.4
+rinoh-typeface-symbola==0.1.1
+Sphinx==5.2.3
```
Comment (jemrobinson) [inline .devcontainer/requirements.txt]: Same with these requirements - they look like a duplicate.

**Review by jemrobinson — CHANGES_REQUESTED** [review]

File: .devcontainer/devcontainer.json [hunk 4]
```diff
@@ -0,0 +1,17 @@
+// Development container for the Data Safe Haven project
+{
+	"name": "Turing Data Safe Haven",
+	"build": {
+		"context": "..",
```
Comment (jemrobinson) [inline .devcontainer/devcontainer.json:5]: This is cool. I don't think Docker can include files from outside parent-of-Dockerfile natively :)

**Review by jemrobinson — APPROVED** [review]

File: .devcontainer/devcontainer.json (hunk 4 above)
Comment (jemrobinson) [reply .devcontainer/devcontainer.json:5]: This is cool. I don't think Docker can include files from outside parent-of-Dockerfile natively :). Merge whenever you're ready @craddm .

**Review by jemrobinson — APPROVED** [review]
This is cool. I don't think Docker can include files from outside parent-of-Dockerfile natively :). Merge whenever you're ready @craddm .

**Comment by craddm** [comment]
I'm not authorized to merge! Yeah, I thought it needed those scripts to be in the same directory (or at least in a sub-directory) as the Dockerfile, but turns out you can just run `docker build` from the parent directory of the `.devcontainer` directory (if you want to build it manually rather than let VSCode do it) and then it handles the filepaths fine.

**Pull request merged**